packages/components/src/dropdown/index.js

```diff
@@ -6,48 +6,45 @@ import classnames from 'classnames';
 /**
  * WordPress dependencies
  */
-import { Component, createRef } from '@wordpress/element';
+import { useRef, useEffect, useState } from '@wordpress/element';
 
 /**
  * Internal dependencies
  */
 import Popover from '../popover';
 
-class Dropdown extends Component {
-	constructor() {
-		super( ...arguments );
+export default function Dropdown( {
+	renderContent,
+	renderToggle,
+	position = 'bottom right',
+	className,
+	contentClassName,
+	expandOnMobile,
+	headerTitle,
+	focusOnMount,
+	popoverProps,
+	onClose,
+	onToggle,
+} ) {
+	const containerRef = useRef();
+	const [ isOpen, setIsOpen ] = useState( false );
+	const isFirstRender = useRef( true );
 
-		this.toggle = this.toggle.bind( this );
-		this.close = this.close.bind( this );
-		this.closeIfFocusOutside = this.closeIfFocusOutside.bind( this );
-
-		this.containerRef = createRef();
-
-		this.state = {
-			isOpen: false,
-		};
-	}
-
-	componentWillUnmount() {
-		const { isOpen } = this.state;
-		const { onToggle } = this.props;
-		if ( isOpen && onToggle ) {
-			onToggle( false );
-		}
-	}
-
-	componentDidUpdate( prevProps, prevState ) {
-		const { isOpen } = this.state;
-		const { onToggle } = this.props;
-		if ( prevState.isOpen !== isOpen && onToggle ) {
+	useEffect( () => {
+		if ( isFirstRender.current ) {
+			isFirstRender.current = false;
+		} else if ( onToggle ) {
 			onToggle( isOpen );
 		}
-	}
+		return () => {
+			if ( isOpen && onToggle ) {
+				onToggle( false );
+			}
+		};
+	}, [ isFirstRender, isOpen, onToggle ] );
```

Comment:
I might be missing an implementation detail, but I think the goal here is to trigger `onToggle` when the state changes, and that could also be handled now as a custom hook, something like the following untested pseudocode:
```jsx
function useObservableState( initialState, onStateChange ) {
    const [ value, setValue ] = useState( initialState );
    return [ value, ( newValue ) => { 
        setValue( newValue ); 
        onStateChange( newValue );
    } ];
}
// ...

const [ isOpen, setIsOpen ] = useIsOpenState( false, onToggle );
```
(I would put this function above the component in the same file)

Would still also need some separate handling for unmounting, but that then becomes simpler:
```jsx
// Fire `onToggle` when the component unmounts
useEffect( () => () => onToggle( false ) );
```

Reply:
Yes. It should work. I'll check it later.